feat: update electricity demand profiles

tgilon opened the pull request:
Closes https://github.com/open-energy-transition/open-tyndp/issues/6.

## Changes proposed in this Pull Request

This PR suggests to add options in PyPSA-Eur to use the custom electricity demand profiles of TYNDP.

The feature can be tested with:
```shell
make tyndp
```

# Tasks
- [x] Archive ENTSO-E data on Zenodo by @daniel-rdt 
- [x] Create switches in the configuration file by @tgilon 
- [x] Implement raw data processing in separate script by @daniel-rdt 
- [x] Check whether `UK` or `GB` should be used
- [x] Define bidding zones as geo data in `build_shapes`
    - [Electricity map topology data](https://github.com/electricitymaps/electricitymaps-contrib/blob/master DATA_SOURCES.md), using https://www.naturalearthdata.com/downloads/10m-cultural-vectors/
    - https://github.com/EnergieID/entsoe-py/tree/master/entsoe/geo/geojson
- [x] Add parameters in `add_electricity` and `prepare_sector_network`by @tgilon 
- [x] Add news rules to documentation
- [x] Address nodes vs countries in electricity demand raw data
- [x] Process data through `build_electricity_demand_base`
- [x] Deactivate Heating technologies (biomass_boilers and oil_boilers) in `prepare_sector_network` `add_biomass`
- [x] Deactivate industrial electricity with new config switch and low temp heat load in `prepare_sector_network` `add_industry` based on `heating` sector switch
- [x] Fix issue with missing Luxembourg `pop_layout` and low-voltage nodes for `LUF1` and `LUG1`
- [x] Fix errors in post processing plotting routines due to missing heat sector
- [x] Check for linear interpolation in `build_electricity_demand`
- [x] Check for manual adjustments in `build_electricity_demand` (should not be allowed)
- [x] Validate `make tyndp`
- [x] Validate that electricity demand in output network corresponds to original data (xref: #44)
- [x] Generalize using `planning_horizons` wildcard

# Workflow
1. `retrieve_tyndp_bundle`
    - Update `outputs` to add electricity demand data
1. (new) `clean_tyndp_data`
    - Prepare the demand data input for `build_electricity_demand` (electricity) and `prepare_sector_network` (other carriers - will be covered later)
3. `build_electricity_demand` and (new) `build_electricity_demand_tyndp`
    - By default, `data/electricity_demand_raw.csv` (OPSD data) and `data/load_synthetic_raw.csv` (demand calculator) are used.
    - `electricity_demand.csv` from PyPSA-Eur is a CSV file with hourly resolution data as rows and countries as columns. TYNDP electricity demand profiles have climate years as columns and tabs for each node. Electricity demand data varies across planning years (hence the need for the `planning_horizons` wildcard for TYNDP demand-related rules).
    - Use of manual adjustments is disabled (as we don't want to alter TYNDP data).
    - Data is given per bidding zones. We want to keep this granularity and will modify the subsequent `build_electricity_demand_base` rule to work with it.
4. `build_electricity_demand_base` and (new) `build_electricity_demand_base_tyndp`
    - Can be skipped for TYNDP data, no upsampling needed
    - Introduction of `planning_horizons` wildcard in the TYNDP case
4. `cluster_network`
    - Disable cluster option in TYNDP case
    - Make `resources/electricity_demand_base_s.nc` an optional input as electrical load is used for clustering
6. `add_electricity`
    - Optional load attached to the network if OPSD load is used
    - If TYNDP load data is used, no load is attached as it relies on `planning_horizons` wildcard. Attached in `prepare_sector_network`.
7. `prepare_sector_network`
    - Electricity
      - Electrical load attached to the network in the TYNDP case (for a given `planning_horizons`).
      - Distribution losses are disabled. TYNDP electricity demand includes T&D losses ([Methodology report](https://2024.entsos-tyndp-scenarios.eu/wp-content/uploads/2024/05/TYNDP_2024_Scenarios_Methodology_Report_240522.pdf), p.21). Therefore, there is no additional modelling of the losses.
    - Transport: 
        - Electric transport demand is treated in https://github.com/open-energy-transition/open-tyndp/issues/5
        - Oil and H2 transport demand can be disabled using shares in the configuration file.
        - Always enable distribution as TYNDP model both prosumer and street nodes
        - Only passenger cars are considered, to be addressed by https://github.com/open-energy-transition/open-tyndp/issues/5
    - Heating: 
        - Considering the fact that most of the heat demand is included in the provided profiles (except HHP), the heating sector is deactivated in the configuration file
        - Electrical heat demand is not removed from the load in the TYNDP case as heat is not explicitly modelled.
        - District heating demand is disabled as it is calculated exogenously using the supply tool.
        - Oil boilers are disabled in the config when using TYNDP load
        - As HHP will not be extended, it makes sense to only add existing in `add_existing_baseyear`. To be addressed by #3
        - Waste heat ? To be addressed by https://github.com/open-energy-transition/open-tyndp/issues/3
    - Biomass:
        - Biomass boilers are disabled in the config when using TYNDP load (see Heating sector).
    - Industry:
        - Shipping and aviation are included over here (e-kerosene, e-diesel)
        - Synthetic fuels modelling is covered in https://github.com/open-energy-transition/open-tyndp/issues/4
        - Low temperatrue industrial heat demand is disabled with the Heating sector switch in TYNDP case as heat is not explicitly modelled. A new dependency is added that ensures low temperature heat industrial heat is only enabled when heating sector is enabled.
        - A new configuration switch `use_industry_load` is introduced for modelling Industrial electricity demand explicitly. This is disabled for TYNDP load.
 8. `add_existing_baseyear`
     - Check existing technologies added
     - Add existing HHP ? Will be addressed by https://github.com/open-energy-transition/open-tyndp/issues/3

## Checklist

- [x] I tested my contribution locally and it works as intended.
- [x] Code and workflow changes are sufficiently documented.
- [ ] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in `config/config.default.yaml`.
- [x] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [x] Changes in configuration options are added in `config/test/*.yaml`.
- [ ] Sources of newly added data are documented in `doc/data_sources.rst`.
- [x] A release note `doc/release_notes.rst` is added.

tgilon commented:
@daniel-rdt Inhouse test fix is merged in upstream: https://github.com/PyPSA/pypsa-eur/pull/1669

tgilon commented on doc/configtables/load.csv:
```diff
@@ -8,3 +9,6 @@ supplement_synthetic,bool,"{true, false}","Whether to supplement missing data fo
 -- population,float,"[0, 1]","Weighting factor for the population data in the distribution key."
+tyndp_scenario,--,--,"Scenario configuration when `tyndp` source is used."
+-- scenario,--,"{"""NT"",""DE"",""GA""}","TYNDP scenario to use as electricity demand, being one of NT, DE or GA."
+-- year,--,"{2030, 2040, 2050}","Scenario year to use, being one of 2030, 2040 or 2050. Electricity demand for 2050 are only defined for DE and GA."
```
`interpolate` argument needs to be added

daniel-rdt replied:
to be addressed by generalized `load["fill_gap"]["enable]` which needs to be added here instead

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -5586,47 +5614,48 @@ def add_industry(
-        carrier="low-temperature heat for industry",
-        p_set=industrial_demand.loc[nodes, "low-temperature heat"] / nhours,
-    )
+    if load_source != "tyndp":
```
This is not strictly necessary to define the electrical demand. But I'm fine to keep it.

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -1648,7 +1656,9 @@ def insert_electricity_distribution_grid(
-    ):
+    ) and "electricity distribution grid" in options["transmission_efficiency"][
+        "enable"
+    ]:
```
If the data we have already includes losses, we probably don't want to model them explicitly. To remove from the `config/config.tyndp.yaml` (and test file).

tgilon replied:
Removed!

tgilon reviewed (commented):
@daniel-rdt Good catch for the industrial electrical load! I deliberately avoided changes to heat as this PR focused on electrical demand. But now it's in, and we need it anyway, so let's keep it.

Building up on this morning's discussion, I propose that this PR inlcudes the validation of the load as a rule.

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -1619,7 +1622,12 @@ def insert_electricity_distribution_grid(
     - Micro-CHP units
     """
-    nodes = pop_layout.index
+
+    nodes = (
+        n.buses.query("carrier == 'AC' and not index.str.contains('DRES')").index
+        if load_source == "tyndp"
+        else pop_layout.index
+    )
```
So ideally we would contribute upstream, generalize this and only add the `DRES` specification in our fork?
```
nodes = n.buses.query("carrier == 'AC' and not index.str.contains('DRES')").index
```

daniel-rdt replied:
Opened a PR on this here: https://github.com/PyPSA/pypsa-eur/pull/1672

tgilon commented:
@daniel-rdt @coroa This PR is ready for me to merge. I used a notebook to check the data. You can find it here: https://github.com/open-energy-transition/open-tyndp/pull/42. This allowed me to catch the fact that GB data was not attached to the network. Could you approve the PR on your side as well?

daniel-rdt reviewed (approved):
Thanks @tgilon! Good catch with the UK / GB load! I also validated the load again on a yearly total load per country basis and it passes the assertion locally. Should we maybe introduce an assertion unit test for this into the workflow based on your validation notebook?

coroa commented on scripts/build_existing_heating_distribution.py:
```diff
@@ -82,7 +82,9 @@ def build_existing_heating():
     pop_layout = pd.read_csv(snakemake.input.clustered_pop_layout, index_col=0)
 
-    nodal_heating = existing_heating.loc[pop_layout.ct]
+    nodal_heating = existing_heating.reindex(pop_layout.ct, fill_value=0).loc[
+        pop_layout.ct.unique()
+    ]
     nodal_heating.index = pop_layout.index
     nodal_heating = nodal_heating.multiply(pop_layout.fraction, axis=0)
```
`
```suggestion
    nodal_heating = existing_heating.reindex(pop_layout.ct, fill_value=0)
    nodal_heating.index = pop_layout.index
    nodal_heating = nodal_heating.multiply(pop_layout.fraction, axis=0)
```

coroa replied:
`.loc[pop_layout.ct.unique()]` is a no-op.

Luckily. Since, if it wasn't, then `nodal_heating.index = pop_layout.index` would fail.

coroa commented on scripts/clean_tyndp_demand.py:
```diff
@@ -0,0 +1,132 @@
+        )
+
+    elif scenario in ["DE", "GA"]:
+        demand_fn = Path(
+            fn,
+            scenario,
+            str(pyear),
+            f"ELECTRICITY_MARKET {scenario} {pyear}.xlsx",
+        )
```
Uh this use of `Path` as a sort of join is new to me, interesting.

tgilon replied:
Ok with keeping it like this?

coroa replied:
Sure, sure. This was a small aha i wanted to share.

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -3565,11 +3578,12 @@ def build_heat_demand(
-        n.loads_t.p_set[electric_nodes]
-        - electric_heat_supply.T.groupby(level=1).sum().T[electric_nodes]
-    )
+    if load_source != "tyndp":
+        electric_nodes = n.loads.index[n.loads.carrier == "electricity"]
+        n.loads_t.p_set[electric_nodes] = (
+            n.loads_t.p_set[electric_nodes]
+            - electric_heat_supply.T.groupby(level=1).sum().T[electric_nodes]
+        )
```
Why don't we have to correct the heat demand by the electric load in the case of the tyndp data?

daniel-rdt replied:
All demand time series (electric and heating are exogenous and already matching each other). We will exchange the heating demand for Hybrid Heat Pumps (HHP) see explanation below, in a dedicated PR. Maybe it is worth separating this a bit more to make this PR neater wrt to heating?

daniel-rdt replied:
@coroa You can also find a concise explanation of this in the PR description above under `Workflow -> Step 7 -> Heating`

coroa replied:
Yes, i noticed after reading through all of the code.

Nothing to change here.

tgilon replied:
Good tks!

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -3688,8 +3706,9 @@ def add_heat(
         solar_thermal = options["solar_cf_correction"] * solar_thermal / 1e3
 
+    heat_systems = HeatSystem if load_source != "tyndp" else []
     for heat_system in (
-        HeatSystem
+        heat_systems
```
Uh, ok, that explains my question above, but is `load_source` which is categorized under the electricity category, really the right configuration to disable all heat demands?

daniel-rdt replied:
The issue is that electricity demand and heat demand in TYNDP are very much related as the electricity input demand time series already includes electricity demand for district heating. Other than this, heating demand is only modelled with so-called Hybrid Heat Pumps (consisting of a gas boilder and an electric heat pump with separate demand time series). Therefore with TYNDP electricity demand included, we thought it to be consistent to already remove the heating technologies.

However, I do agree that it is incomplete here (as this PR was.not meant cover implementing all the heat demand and technologies).

In terms of configuration, I guess that a general refactoring to pool all the TYNDP settings in one section would help. I agree though, that we should probably use a separate configuration for TYNDP heating demands.

tgilon replied:
As @daniel-rdt said, I used the electricity demand source here because it includes electrical heat demand. The profiles for this demand computed by the DFT tool and are exogenous. This means that we can safely disable this section when relying on this electrical heat demand.

However, I understand that we might want to generalise this a bit (and contribute upstream?). In this case, a standalone switch would make sense. I would avoid it if it's not strictly necessary, bearing in mind the need to keep the tyndp config as simple as possible.

tgilon replied:
See https://github.com/open-energy-transition/open-tyndp/pull/14#issuecomment-2880100511

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -4885,7 +4907,7 @@ def add_biomass(
         )
 
-    if options["biomass_boiler"]:
+    if options["biomass_boiler"] and load_source != "tyndp":
```
same

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -5373,8 +5398,8 @@ def add_industry(
         nodes = pop_layout.index
-
-        for heat_system in HeatSystem:
+        heat_systems = HeatSystem if load_source != "tyndp" else []
+        for heat_system in heat_systems:
```
same

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -5586,47 +5611,48 @@ def add_industry(
-        carrier="low-temperature heat for industry",
-        p_set=industrial_demand.loc[nodes, "low-temperature heat"] / nhours,
-    )
+    if load_source != "tyndp":
```
same as above?

tgilon replied:
Added a new config `sector:use_industry_load`

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -7180,7 +7222,19 @@ def add_import_options(
+        and snakemake.params.load_source == "tyndp"
+    ):
+        logger.warning(
+            "Distribution network should always be used when using TYNDP electric load."
```
Can you remind me why?

tgilon replied:
The TYNDP models EV with both street and home charger. It's not a must, but I thought it would serve as a good reminder for the user.

coroa reviewed (commented):
Looks good in general, a few small improvements and questions, i do still have :)

coroa commented:
Re: Heating demand. It would be better if we can deactivate the heating demand with some heating specific settings (since this can then also avoid calculating all the heat pump profiles and temperature dependent heat demand inputs).

coroa commented:
Maybe we need to add a sort of validate config step during which incompatible settings are raising.

tgilon commented:
> Should we maybe introduce an assertion unit test for this into the workflow based on your validation notebook?

I thought about it. Since we need the output of `prepare_sector_network`, I haven't been able to identify a useful test.

tgilon commented on .github/pull_request_template.md:
```diff
@@ -10,5 +10,6 @@ Closes # (if applicable).
 - [ ] Changed dependencies are added to `envs/environment.yaml`.
 - [ ] Changes in configuration options are added in `config/config.default.yaml`.
 - [ ] Changes in configuration options are documented in `doc/configtables/*.csv`.
+- [ ] Changes in configuration options are added in `config/test/*.yaml`.
```
I'm aware of the following https://github.com/open-energy-transition/open-tyndp/pull/14/commits/c6444affed24287b45f51402736942c11fb60458. Expect this fix, it is original work.

coroa replied:
Thanks

daniel-rdt replied:
The changes from https://github.com/open-energy-transition/open-tyndp/commit/c6444affed24287b45f51402736942c11fb60458 are also already in the [OET fork](https://github.com/open-energy-transition/pypsa-eur/commits/master/) and on the [`open-tyndp` master ](https://github.com/open-energy-transition/open-tyndp/commits/master/)

tgilon commented on config/config.default.yaml:
```diff
@@ -337,16 +337,19 @@ transformers:
     gdp: 0.6
     population: 0.4
-
+  tyndp_scenario: DE  # NT, DE or GA
```
I suggest setting an empty scenario as the default option. My current approach is to create a TYNDP config file (or a set of scenarios, let's see). Once I have a better idea of the set of configuration required, I would suggest converging on a single TYNDP-enable config, if possible.

daniel-rdt replied:
fyi @coroa, we also already have an open issue related to this refactoring (https://github.com/open-energy-transition/open-tyndp/issues/39)

tgilon commented on config/config.tyndp.yaml:
```diff
@@ -26,13 +27,13 @@ enable:
 
 co2_budget:
   # TODO define carbon budget compliant with EU 2030 and 2050 targets
```
Nope, this simply aligns the CO2 budget with the updated default.

tgilon commented on config/test/config.tyndp.yaml:
```diff
@@ -3,20 +3,21 @@
   disable_progressbar: true
   shared_resources:
-    policy: "test"
+    policy: false
```
This is the new default since https://github.com/PyPSA/pypsa-eur/commit/037b196104b2ea77509a1d05e1aef72c58d79bf4.

coroa replied:
Ok, fine. Carried that question to the PR there.

daniel-rdt replied:
My understanding is that the shared_resources don't work currently on the upstream master (see issue https://github.com/PyPSA/pypsa-eur/issues/1576)

tgilon commented on config/test/config.tyndp.yaml:
```diff
@@ -27,6 +28,10 @@ snapshots:
 
+co2_budget:
+  # TODO define carbon budget compliant with EU 2030 and 2050 targets
+  2040: 0.2  # Temporary increase to allow CI to pass while load configurations are incomplete.
```
I think that we need to implement the H2 config before reverting this.

coroa replied:
Fine

tgilon commented on scripts/add_existing_baseyear.py:
```diff
@@ -511,6 +512,8 @@ def add_heating_capacities_installed_before_baseyear(
         Minimum capacity threshold
     use_electricity_distribution_grid : bool
         Whether to use electricity distribution grid
+    load_source : str (optional)
```
My bad, it's consistent expect in OSM related scripts (e.g. https://github.com/PyPSA/pypsa-eur/blob/master/scripts/build_osm_network.py#L134-L150)

daniel-rdt replied:
I noted down to check all the OSM related scripts for numpy convention and refactor. We noticed this before and would be good to contribute upstream

tgilon commented on scripts/build_electricity_demand.py:
```diff
@@ -270,14 +320,23 @@ def manual_adjustment(load, fn_load, countries):
+        logger.info(
+            f"Filling larger gaps by copying time-slices of period '{time_shift}'."
+        )
+        load = load.apply(fill_large_gaps, shift=time_shift)
```
Sure

tgilon commented on scripts/build_electricity_demand.py:
```diff
@@ -20,33 +20,70 @@
     fn : str
         File name or url location (file format .csv)
+    years : None or slice()
+        Years for which to read load data (defaults to slice("2018","2019"))
```
Good catch! I changed the implementation slightly and forgot about the doc.

I'll change it to
```python
    years : slice(pandas.DatetimeIndex)
        Years for which to read load data
```

tgilon commented on scripts/build_electricity_demand.py:
```diff
@@ -232,7 +269,10 @@ def manual_adjustment(load, fn_load, countries):
         from scripts._helpers import mock_snakemake
 
-        snakemake = mock_snakemake("build_electricity_demand")
+        snakemake = mock_snakemake(
+            "build_electricity_demand",
+            planning_horizons=2030,
+        )
```
Sure

coroa commented:
So, nothing here is blocking anymore in my opinion.

I would like that the `load_source != "tyndp"` fences vanish in favour of existing or new heat sector config settings, which at the moment should be only documented to be co-dependent.

And before the final tyndp release we are aiming for some sort of config validation, which complains if incompatible settings are chosen.

tgilon commented:
@coroa @daniel-rdt Here is a summary of what I did:
- `biomass_boilers` and `oil_boilers` are existing config. I've disabled them.
- I've removed the warning for the electricity distribution grid as this will be addressed by the config validation.
- I introduced a new config for the industrial load: `sector:use_industry_load`. This is similar to the heating question.

All of this is documented in https://github.com/open-energy-transition/open-tyndp/issues/43.
